### PR TITLE
Fix drupal_set_title() changing breadcrumb titles

### PR DESCRIPTION
**Pull request by darrenoh**

### CODE REVIEW

**Review by mglaman — COMMENTED** [review]
What's a way to reproduce this? I wonder if we could make `RetrofitTitleResolver` aware of the `RequestStack` instead. Then we can use `SplObjectStorage` to store the title based on objects.

That is how we handled it in Commerce for request-specific things https://git.drupalcode.org/project/commerce/-/blob/8.x-2.x/modules/store/src/CurrentStore.php?ref_type=heads

**Comment by darrenoh** [comment]
I experienced this with Ctools on export UI page callbacks. Each callback ran drupal_set_title(), so when the breadcrumb was generated every link had the same title.

Object storage sounds good, but I don't know what RequestStack is.

**Comment by mglaman** [comment]
The RequestStack is how `\Drupal::request()` works, it tracks all the requests being handled. The `RetrofitTitleResolver` would actually look like the `path.current` service I think: https://git.drupalcode.org/project/drupal/-/blob/11.x/core/lib/Drupal/Core/Path/CurrentPathStack.php?ref_type=heads

File: tests/src/Integration/RetrofitTitleResolverTest.php [hunk 1]
```diff
@@ -45,13 +46,17 @@ public function testSetStoredTitle(): void
             RetrofitTitleResolver::class,
             $titleResolver
         );
+        $requestStack = $this->container->get('request_stack');
+        assert($requestStack instanceof RequestStack);
+        $request = $requestStack->getCurrentRequest();
+        assert($request instanceof Request);
```
Comment (mglaman) [inline tests/src/Integration/RetrofitTitleResolverTest.php:52]: Huh, I'm surprised this works without pushing a request to the stack.

**Review by mglaman — APPROVED** [review]
Looks great!

File: tests/src/Integration/RetrofitTitleResolverTest.php (hunk 1 above)
Comment (darrenoh) [reply tests/src/Integration/RetrofitTitleResolverTest.php:52]: I found that `pathInfo` is not set when a request is created, so it never matched the request from the request stack.

**Pull request merged**